tests/MongoDB.Driver.Core.Tests/Core/Operations/AsyncCursorTests.cs

```diff
@@ -313,143 +311,151 @@ private ConnectionDescription CreateConnectionDescriptionSupportingSession()
                 new MessageEncoderSettings(),
                 maxTime.WithDefault(null));
         }
+    }
 
-        // nested types
-        private class Reflector
+    public class AsyncCursorIntegrationTests : OperationTestBase
+    {
+        [Theory]
+        [InlineData(0, 1000)]
+        [InlineData(2, 2)]
+        [InlineData(2, 1000)]
+        [InlineData(4, 2)]
+        [InlineData(4, 4)]
+        [InlineData(4, 1000)]
+        public void Session_reference_count_should_be_decremented_as_soon_as_possible(int collectionSize, int batchSize)
         {
-            // private fields
-            private AsyncCursor<BsonDocument> _instance;
+            RequireServer.Check();
+            DropCollection();
+            var documents = Enumerable.Range(1, collectionSize).Select(n => new BsonDocument("_id", n));
+            Insert(documents);
 
-            // constructors
-            public Reflector(AsyncCursor<BsonDocument> instance)
+            var findOperation = new FindOperation<BsonDocument>(_collectionNamespace, BsonDocumentSerializer.Instance, _messageEncoderSettings)
```

Comment:
Try writing this test without relying on FindOperation.  See [QueryBatchCursorFunctionalSpecification.groovy](https://github.com/mongodb/mongo-java-driver/blob/master/driver-core/src/test/functional/com/mongodb/operation/QueryBatchCursorFunctionalSpecification.groovy) for how java driver does it.  It has the advantage of giving access to the ChannelSource in all cases, so that you can actually assert that it's been disposed of.